dhctl/pkg/config/deckhouse_config.go

```diff
@@ -0,0 +1,257 @@
+/*
+Copyright 2023 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/Masterminds/semver/v3"
+
+	"github.com/deckhouse/deckhouse/dhctl/pkg/app"
+	"github.com/deckhouse/deckhouse/dhctl/pkg/log"
+)
+
+const (
+	initConfigurationError = `%s fields in InitConfiguration are deprecated.
+Please use ModuleConfig 'deckhouse' section in configuration. Example:
+---
+apiVersion: deckhouse.io/v1alpha1
+kind: ClusterConfiguration
+...
+apiVersion: deckhouse.io/v1alpha1
+kind: InitConfiguration
+...
+---
+apiVersion: deckhouse.io/v1alpha1
+kind: ModuleConfig
+metadata:
+  name: deckhouse
+spec:
+  settings:
+    %s
+`
+)
+
+type DeckhouseInstaller struct {
+	Registry              RegistryData
+	LogLevel              string
+	Bundle                string
+	DevBranch             string
+	UUID                  string
+	KubeDNSAddress        string
+	ClusterConfig         []byte
+	ProviderClusterConfig []byte
+	StaticClusterConfig   []byte
+	TerraformState        []byte
+	NodesTerraformState   map[string][]byte
+	CloudDiscovery        []byte
+	ModuleConfigs         []*ModuleConfig
+
+	KubeadmBootstrap   bool
+	MasterNodeSelector bool
+
+	ReleaseChannel string
+}
+
+func (c *DeckhouseInstaller) GetImage(forceVersionTag bool) string {
+	registryNameTemplate := "%s%s:%s"
+	tag := c.DevBranch
+	if forceVersionTag {
+		versionTag, foundValidTag := readVersionTagFromInstallerContainer()
+		if foundValidTag {
+			tag = versionTag
+		}
+	}
+
+	if tag == "" {
+		panic("Probably you use development image. please use devBranch")
+	}
+
+	return fmt.Sprintf(registryNameTemplate, c.Registry.Address, c.Registry.Path, tag)
+}
+
+func (c *DeckhouseInstaller) IsRegistryAccessRequired() bool {
+	return c.Registry.DockerCfg != ""
+}
+
+func readVersionTagFromInstallerContainer() (string, bool) {
+	rawFile, err := os.ReadFile(app.VersionFile)
+	if err != nil {
+		log.WarnF(
+			"Could not read %s: %v\nWill fall back to installation from release channel or dev branch.",
+			app.VersionFile, err,
+		)
+		return "", false
+	}
+
+	tag := string(rawFile)
+	if _, err = semver.NewVersion(tag); err != nil {
+		return "", false
+	}
+
+	return tag, true
+}
+
+func PrepareDeckhouseInstallConfig(metaConfig *MetaConfig) (*DeckhouseInstaller, error) {
```

Comment:
This function is very long and has high cognitive complexity, there should be some way to split it

Reply:
Like, at least the part where you handle deckhouse ModuleConfig does not belong here and can be extracted into a separate method or function.